llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp

```diff
@@ -2555,12 +2555,33 @@ bool SIRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator MI,
                       TmpResultReg)
                   .addImm(ST.getWavefrontSizeLog2())
                   .addReg(FrameReg);
-              auto Add = BuildMI(*MBB, MI, DL, TII->get(AMDGPU::V_ADD_U32_e32),
-                                 TmpResultReg);
-              Add.addImm(Offset).addReg(TmpResultReg, RegState::Kill);
+
+              MachineInstrBuilder Add;
+              if ((Add = TII->getAddNoCarry(*MBB, MI, DL, TmpResultReg, *RS)) ==
+                  nullptr) {
+                // VCC is live and no SGPR is free.
+                // since emergency stack slot is already used for spilling VGPR
+                // scavenged? This a way around to avoid carry, need follow-up.
+                BuildMI(*MBB, MI, DL, TII->get(AMDGPU::S_MOV_B32), ResultReg)
+                    .addImm(Offset);
+                Add = BuildMI(*MBB, MI, DL, TII->get(AMDGPU::V_MAD_I32_I24_e64),
```

Comment:
You could handle using mad like this directly in getAddNoCarry.

I was also originally thinking of folding in the FP scaling into the mad. I.e. v_mad_u32_u24 fp, wavesize, offset

Reply:
Then i may i have to modify the interface for getAddNoCarry, as we are using an scavenged register for shifting Offset into.
or that part should i take care as present?

Reply:
I guess it's split to avoid the scavenging inside. I guess just reorder the handling for now

Reply:
But it's going to shift around if we can fold in the shift by wavesize

Reply:
There are more cases need to be handled if i move handling inside, getAddNoCarry.
It hits assert(MIB->getOpcode() == AMDGPU::V_ADD_CO_U32_e64 && "Need to reuse carry out register");
where we are trying to handle !SALU cases.

for now I have left it custom handled.